qutebrowser/browser/history.py

```diff
@@ -283,14 +162,49 @@ def add_url(self, url, title="", *, redirect=False, atime=None):
         if atime is None:
             atime = time.time()
         entry = Entry(atime, url, title, redirect=redirect)
-        if self._initial_read_done:
-            self._add_entry(entry)
-            self._new_history.append(entry)
-            self.item_added.emit(entry)
-            if not entry.redirect:
-                self.add_completion_item.emit(entry)
+        self._add_entry(entry)
+
+    def _parse_entry(self, line):
+        """Parse a history line like '12345 http://example.com title'."""
+        data = line.split(maxsplit=2)
+        if len(data) == 2:
+            atime, url = data
+            title = ""
+        elif len(data) == 3:
+            atime, url, title = data
         else:
-            self._add_entry(entry, target=self._temp_history)
+            raise ValueError("2 or 3 fields expected")
+
+        url = QUrl(url)
+        if not url.isValid():
+            raise ValueError("Invalid URL: {}".format(url.errorString()))
+
+        # https://github.com/qutebrowser/qutebrowser/issues/670
+        atime = atime.lstrip('\0')
+
+        if '-' in atime:
+            atime, flags = atime.split('-')
+        else:
+            flags = ''
+
+        if not set(flags).issubset('r'):
+            raise ValueError("Invalid flags {!r}".format(flags))
+
+        redirect = 'r' in flags
+
+        return (url, title, float(atime), bool(redirect))
+
+    def read(self, path):
+        """Import a text file into the sql database."""
+        with open(path, 'r') as f:
+            rows = []
+            for line in f:
+                try:
+                    row = self._parse_entry(line.strip())
+                    rows.append(row)
+                except ValueError:
+                    log.init.warning('Skipping history line {}'.format(line))
```

Comment:
Makes sense, but I don't want everyone to have qutebrowser break the next time they open it. My personal history file has several blank lines in it, so it would break for me. I don't know how they got there, but I've been happily proceeding without knowing because the previous behavior was to ignore them.